components/popover/index.js

```diff
@@ -133,17 +133,27 @@ class Popover extends Component {
 	}
 
 	setOffset() {
+		const { range } = this.props;
 		const { anchor, popover } = this.nodes;
 		if ( ! anchor || ! anchor.parentNode ) {
 			return;
 		}
 
-		const rect = anchor.parentNode.getBoundingClientRect();
 		const [ yAxis, xAxis ] = this.getPositions();
 		const isTop = 'top' === yAxis;
 		const isLeft = 'left' === xAxis;
 		const isRight = 'right' === xAxis;
-
+		let rect = anchor.parentNode.getBoundingClientRect();
+		if ( range ) {
```

Comment:
Is the general idea that if we pass a range, we use it as a anchor reference instead of `this.nodes.anchor` ? If so, do we need to do anything else about the logic below for padding offset (should this be eliminated)?

Or, to my earlier points about making Popover unaware of range, would one of the following options work:

- Allow parent component to pass an override for setting offset, like `onSetOffset` or `getOffsetRect` ?
- Allow parent component to pass an anchor object which implements `getBoundingClientRect` (`Range`, `Element`)?

Reply:
> Is the general idea that if we pass a range, we use it as a anchor reference instead of this.nodes.anchor ?

yes

> If so, do we need to do anything else about the logic below for padding offset (should this be eliminated)?

Yes it looks like it is not needed when positioned over a range.

> Or, to my earlier points about making Popover unaware of range, would one of the following options work:
> * Allow parent component to pass an override for setting offset, like onSetOffset or getOffsetRect ?
> * Allow parent component to pass an anchor object which implements getBoundingClientRect (Range, Element)?

They'd probably both work to some degree though the second option would be problematic when ranges wrapped across multiple lines.

Reply:
I have changed the calculation of the topOffset so it only happens for the anchor.

Reply:
I have extracted the getBoundingClientRect logic into a function getAnchorRect which can be passed to override the default of positioning around the parent element.

So it now behaves like your first suggested option.